ruspy/simulation/simulation_auxiliary.py

```diff
@@ -213,22 +50,56 @@ def simulate_strategy_loop_known(
         :num_states: (int) : The size of the state space.
 
     """
+    num_states = ev.shape[0]
+    num_periods = decisions.shape[1]
     for period in range(num_periods):
-        for bus in range(num_buses):
-            old_state = states[bus, period]
-            if (-costs[old_state, 0] + unobs[bus, period, 0] + beta * ev[old_state]) > (
-                -costs[0, 0] - costs[0, 1] + unobs[bus, period, 1] + beta * ev[0]
-            ):
-                decision = 0
-                utility = -costs[old_state, 0] + unobs[bus, period, 0]
-                new_state = old_state + increments[bus, period]
-            else:
-                decision = 1
-                utility = -costs[0, 0] - costs[0, 1] + unobs[bus, period, 1]
-                new_state = increments[bus, period]
-
-            decisions[bus, period] = decision
-            utilities[bus, period] = utility
-            if period < num_periods - 1:
-                states[bus, period + 1] = new_state
+        old_state = states[bus, period]
+        if (-costs[old_state, 0] + unobs[bus, period, 0] + beta * ev[old_state]) > (
+            -costs[0, 0] - costs[0, 1] + unobs[bus, period, 1] + beta * ev[0]
+        ):
+            decision = 0
+            utility = -costs[old_state, 0] + unobs[bus, period, 0]
+            new_state = old_state + increments[old_state, period]
+        else:
+            decision = 1
+            utility = -costs[0, 0] - costs[0, 1] + unobs[bus, period, 1]
+            new_state = increments[0, period]
+
+        decisions[bus, period] = decision
+        utilities[bus, period] = utility
+        if period < num_periods - 1:
+            states[bus, period + 1] = new_state
+        if new_state > num_states - 10:
+            raise ValueError("State space is too small.")
     return states, decisions, utilities
+
+
+def get_unobs(shock, num_buses, num_periods):
```

Comment:
See earlier comment on structure. This just looks way too complicated for what I suspect it is doing.